metrics/metrics.go

```diff
@@ -0,0 +1,194 @@
+	cmm := csiMetricsManager{
+		registry: metrics.NewKubeRegistry(),
+		csiOperationsLatencyMetric: metrics.NewHistogramVec(
+			&metrics.HistogramOpts{
+				Subsystem:      subsystem,
+				Name:           operationsLatencyMetricName,
+				Help:           operationsLatencyHelp,
+				Buckets:        operationsLatencyBuckets,
+				StabilityLevel: metrics.ALPHA,
+			},
+			[]string{labelCSIDriverName, labelCSIOperationName, labelGrpcStatusCode},
+		),
+	}
+
+	cmm.SetDriverName(driverName)
+	cmm.registerMetrics()
+	return &cmm
+}
+
+var _ CSIMetricsManager = &csiMetricsManager{}
+
+type csiMetricsManager struct {
+	registry                   metrics.KubeRegistry
+	driverName                 string
+	csiOperationsMetric        *metrics.CounterVec
+	csiOperationsLatencyMetric *metrics.HistogramVec
+}
+
+func (cmm *csiMetricsManager) GetRegistry() metrics.KubeRegistry {
+	return cmm.registry
+}
+
+// RecordMetrics must be called upon CSI Operation completion to record
+// the operation's metric.
+// operationName - Name of the CSI operation.
+// operationErr - Error, if any, that resulted from execution of operation.
+// operationDuration - time it took for the operation to complete
+func (cmm *csiMetricsManager) RecordMetrics(
+	operationName string,
+	operationErr error,
+	operationDuration time.Duration) {
+	cmm.csiOperationsLatencyMetric.WithLabelValues(
+		cmm.driverName, operationName, getErrorCode(operationErr)).Observe(operationDuration.Seconds())
+}
+
+// SetDriverName is called to update the CSI driver name. This should be done
+// as soon as possible, otherwise metrics recorded by this manager will be
+// recorded with an "unknown-driver" driver_name.
+func (cmm *csiMetricsManager) SetDriverName(driverName string) {
+	if driverName == "" {
+		cmm.driverName = unknownCSIDriverName
+	} else {
+		cmm.driverName = driverName
+	}
+}
+
+// StartMetricsEndpoint starts the metrics endpoint at the specified address/path
+// for this metrics manager  on a new go routine.
+// If the metricsAddress is an empty string, this will be a no op.
+func (cmm *csiMetricsManager) StartMetricsEndpoint(metricsAddress, metricsPath string) {
+	if metricsAddress == "" {
+		klog.Warningf("metrics endpoint will not be started because `metrics-address` was not specified.")
+		return
+	}
+
+	http.Handle(metricsPath, metrics.HandlerFor(
+		cmm.GetRegistry(),
+		metrics.HandlerOpts{
+			ErrorHandling: metrics.ContinueOnError}))
+
+	// Spawn a new go routine to listen on specified endpoint
+	go func() {
+		err := http.ListenAndServe(metricsAddress, nil)
+		if err != nil {
+			klog.Fatalf("Failed to start prometheus metrics endpoint on specified address (%q) and path (%q): %s", metricsAddress, metricsPath, err)
+		}
+	}()
+}
+
+// VerifyMetricsMatch is a helper function that verifies that the expected and
+// actual metrics are identical excluding metricToIgnore.
+func VerifyMetricsMatch(expectedMetrics, actualMetrics string, metricToIgnore string) error {
```

Comment:
This is only used in test right? Can we move it to the test file instead of exporting it?

Reply:
This method is only used by tests. Ideally it should be in the _test file,  but *_test.go files are compiled into the package only when running go test for that package and this method is used by metrics_test as well as connection_test. In the future, if there are more consumers, we can consider moving it to a new, standalone package. For now, I added a comment documenting this.